Create sequential serializer failed

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-fury/issues) and found no similar issues.

It may be a duplicate of https://github.com/apache/incubator-fury/issues/1176, but the trace suggests it **might** be a different cause.


### Version

0.4.1

### Component(s)

Java

### Minimal reproduce step

Run the serializer against our object graph. It seems to happen intermittently, so could be a threading issue. 

### What did you expect to see?

No runtime exception.

### What did you see instead?

```
17:02:55  Exception in thread "fury-jit-compiler-2" java.lang.RuntimeException: Create sequential serializer failed, 
class: class io.citrine.lolo.bags.RegressionBaggerTrainingResult
17:02:55  	at io.fury.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:48)
17:02:55  	at io.fury.resolver.ClassResolver.lambda$getObjectSerializerClass$2(ClassResolver.java:963)
17:02:55  	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
17:02:55  	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
17:02:55  	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
17:02:55  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
17:02:55  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
17:02:55  	at java.base/java.lang.Thread.run(Thread.java:829)
17:02:55  Caused by: java.lang.IllegalArgumentException: Expected AbstractCollectionSerializer but got io.fury.serializer.Serializer
17:02:55  	at io.fury.util.Preconditions.checkArgument(Preconditions.java:78)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeForCollection(BaseObjectCodecBuilder.java:1210)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeForNotNull(BaseObjectCodecBuilder.java:1162)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.lambda$readContainerElement$12(BaseObjectCodecBuilder.java:1401)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.readRef(BaseObjectCodecBuilder.java:1093)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.readContainerElement(BaseObjectCodecBuilder.java:1398)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.lambda$readContainerElements$2223955c$1(BaseObjectCodecBuilder.java:1360)
17:02:55  	at io.fury.codegen.Expression$ForLoop.doGenCode(Expression.java:2363)
17:02:55  	at io.fury.codegen.Expression.genCode(Expression.java:102)
17:02:55  	at io.fury.codegen.Expression$ListExpression.doGenCode(Expression.java:181)
17:02:55  	at io.fury.codegen.Expression.genCode(Expression.java:102)
17:02:55  	at io.fury.codegen.ExpressionOptimizer.invokeGenerated(ExpressionOptimizer.java:120)
17:02:55  	at io.fury.codegen.ExpressionOptimizer.invokeGenerated(ExpressionOptimizer.java:68)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.readCollectionCodegen(BaseObjectCodecBuilder.java:1298)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeForCollection(BaseObjectCodecBuilder.java:1220)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeForNotNull(BaseObjectCodecBuilder.java:1162)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeForNotNull(BaseObjectCodecBuilder.java:1122)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.lambda$deserializeFor$7(BaseObjectCodecBuilder.java:1073)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.readRef(BaseObjectCodecBuilder.java:1093)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeFor(BaseObjectCodecBuilder.java:1073)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.deserializeFor(BaseObjectCodecBuilder.java:1057)
17:02:55  	at io.fury.builder.ObjectCodecBuilder.lambda$deserializeGroup$37fcf467$1(ObjectCodecBuilder.java:522)
17:02:55  	at io.fury.codegen.ExpressionOptimizer.invokeGenerated(ExpressionOptimizer.java:48)
17:02:55  	at io.fury.builder.ObjectCodecOptimizer.invokeGenerated(ObjectCodecOptimizer.java:134)
17:02:55  	at io.fury.builder.ObjectCodecBuilder.deserializeGroup(ObjectCodecBuilder.java:538)
17:02:55  	at io.fury.builder.ObjectCodecBuilder.buildDecodeExpression(ObjectCodecBuilder.java:439)
17:02:55  	at io.fury.builder.BaseObjectCodecBuilder.genCode(BaseObjectCodecBuilder.java:204)
17:02:55  	at io.fury.codegen.CompileUnit.getCode(CompileUnit.java:54)
17:02:55  	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:72)
17:02:55  	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:64)
17:02:55  	at io.fury.codegen.CodeGenerator.compile(CodeGenerator.java:144)
17:02:55  	at io.fury.builder.CodecUtils.loadOrGenCodecClass(CodecUtils.java:91)
17:02:55  	at io.fury.builder.CodecUtils.loadOrGenObjectCodecClass(CodecUtils.java:42)
17:02:55  	at io.fury.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:45)
17:02:55  	... 7 more
```

### Anything Else?

I might be doing something wrong, but would like to help to triage.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/org/apache/fury/builder/BaseObjectCodecBuilder.java
@@ -628,23 +628,13 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
   }
 
   protected TypeToken<?> getSerializerType(TypeToken<?> objType) {
-    if (COLLECTION_TYPE.isSupertypeOf(objType)) {
+    Class<?> rawType = getRawType(objType);
+    if (classResolver.isCollection(rawType)) {
       return COLLECTION_SERIALIZER_TYPE;
-    } else if (MAP_TYPE.isSupertypeOf(objType)) {
+    } else if (classResolver.isMap(rawType)) {
       return MAP_SERIALIZER_TYPE;
     }
     return SERIALIZER_TYPE;
-  }
-
-  protected Expression castSerializer(Expression serializer, TypeToken<?> objType) {
-    if (COLLECTION_TYPE.isSupertypeOf(objType)
-        && !COLLECTION_SERIALIZER_TYPE.isSupertypeOf(serializer.type())) {
-      serializer = new Cast(serializer, COLLECTION_SERIALIZER_TYPE, "colSerializer");
-    } else if (MAP_TYPE.isSupertypeOf(objType)
-        && !MAP_SERIALIZER_TYPE.isSupertypeOf(serializer.type())) {
-      serializer = new Cast(serializer, TypeToken.of(AbstractMapSerializer.class), "mapSerializer");
-    }
-    return serializer;
   }
 
   /**
@@ -751,16 +741,17 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
           neq(new BitAnd(flags, notDeclTypeFlag), notDeclTypeFlag, "isDeclType");
       Expression elemSerializer; // make it in scope of `if(sameElementClass)`
       boolean maybeDecl = visitFury(f -> f.getClassResolver().isSerializable(elemClass));
+      TypeToken<?> serializerType = getSerializerType(elementType);
       if (maybeDecl) {
         elemSerializer =
             new If(
                 isDeclType,
-                getOrCreateSerializer(elemClass),
-                castSerializer(writeElementsHeader.f1.inline(), elementType),
+                new Cast(getOrCreateSerializer(elemClass), serializerType),
+                new Cast(writeElementsHeader.f1.inline(), serializerType),
                 false,
-                getSerializerType(elementType));
+                serializerType);
       } else {
-        elemSerializer = castSerializer(writeElementsHeader.f1.inline(), elementType);
+        elemSerializer = new Cast(writeElementsHeader.f1.inline(), serializerType);
       }
       elemSerializer = uninline(elemSerializer);
       Expression action;
@@ -1272,12 +1263,12 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
         elemSerializer =
             new If(
                 isDeclType,
-                getOrCreateSerializer(elemClass),
-                castSerializer(serializer.inline(), elementType),
+              new Cast(getOrCreateSerializer(elemClass), serializerType),
+              new Cast(serializer.inline(), serializerType),
                 false,
                 serializerType);
       } else {
-        elemSerializer = castSerializer(serializer.inline(), elementType);
+        elemSerializer = new Cast(serializer.inline(), serializerType);
       }
       elemSerializer = uninline(elemSerializer);
       builder.add(sameElementClass);

--- a/java/fury-core/src/main/java/org/apache/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/org/apache/fury/builder/BaseObjectCodecBuilder.java
@@ -1263,8 +1263,8 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
         elemSerializer =
             new If(
                 isDeclType,
-              new Cast(getOrCreateSerializer(elemClass), serializerType),
-              new Cast(serializer.inline(), serializerType),
+                new Cast(getOrCreateSerializer(elemClass), serializerType),
+                new Cast(serializer.inline(), serializerType),
                 false,
                 serializerType);
       } else {

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -111,11 +111,11 @@ public class CollectionSerializersTest extends FuryTestBase {
   @Test(dataProvider = "referenceTrackingConfig")
   public void testBasicListNestedJIT(boolean referenceTracking) {
     Fury fury =
-      Fury.builder()
-        .withRefTracking(referenceTracking)
-        .withCodegen(true)
-        .requireClassRegistration(false)
-        .build();
+        Fury.builder()
+            .withRefTracking(referenceTracking)
+            .withCodegen(true)
+            .requireClassRegistration(false)
+            .build();
     List<List<List<String>>> list = new ArrayList<>();
     list.add(ofArrayList(ofArrayList("a", "b")));
     list.add(ofArrayList(ofArrayList("a", "b")));

--- a/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/serializer/collection/CollectionSerializersTest.java
@@ -21,6 +21,7 @@ package org.apache.fury.serializer.collection;
 
 import static org.apache.fury.collection.Collections.ofArrayList;
 import static org.apache.fury.collection.Collections.ofHashMap;
+import static org.apache.fury.collection.Collections.ofHashSet;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertSame;
 
@@ -97,6 +98,33 @@ public class CollectionSerializersTest extends FuryTestBase {
     serDeCheckSerializer(fury, Arrays.asList("a", "b", "c"), "ArraysAsList");
     serDeCheckSerializer(fury, new HashSet<>(data), "HashSet");
     serDeCheckSerializer(fury, new LinkedHashSet<>(data), "LinkedHashSet");
+  }
+
+  @Data
+  public static class BasicListNestedJIT {
+    public Set<List<List<String>>> data;
+    public Set<Set<List<String>>> data1;
+    public Collection<ArrayList<Object>> data2;
+    public Collection<Set<Collection<String>>> data3;
+  }
+
+  @Test(dataProvider = "referenceTrackingConfig")
+  public void testBasicListNestedJIT(boolean referenceTracking) {
+    Fury fury =
+      Fury.builder()
+        .withRefTracking(referenceTracking)
+        .withCodegen(true)
+        .requireClassRegistration(false)
+        .build();
+    List<List<List<String>>> list = new ArrayList<>();
+    list.add(ofArrayList(ofArrayList("a", "b")));
+    list.add(ofArrayList(ofArrayList("a", "b")));
+    BasicListNestedJIT o = new BasicListNestedJIT();
+    o.data = new HashSet<>(list);
+    o.data1 = ofHashSet(ofHashSet(ofArrayList("a", "b")), ofHashSet(ofArrayList("a", "b")));
+    o.data2 = ofHashSet(ofArrayList("a", "b"));
+    o.data3 = ofHashSet(ofHashSet(ofArrayList("a", "b")), ofHashSet(ofArrayList("a", "b")));
+    serDeCheckSerializer(fury, o, "Codec");
   }
 
   @Test(dataProvider = "referenceTrackingConfig")

--- a/scala/src/test/scala/org/apache/fury/serializer/CollectionSerializerTest.scala
+++ b/scala/src/test/scala/org/apache/fury/serializer/CollectionSerializerTest.scala
@@ -63,6 +63,10 @@ class CollectionSerializerTest extends AnyWordSpec with Matchers {
         val struct2 = CollectionStruct1(Nil)
         fury1.deserialize(fury1.serialize(struct2)) shouldEqual struct2
       }
+      "serialize/deserialize NestedCollectionStruct" in {
+        val struct = NestedCollectionStruct(List(List("a", "b"), List("a", "b")), Set(Set("c", "d")))
+        fury1.deserialize(fury1.serialize(struct)) shouldEqual struct
+      }
     }
     s"fury scala map support: setOpt $setOpt, setFactory $setFactory" should {
       "serialize/deserialize Map" in {
@@ -77,10 +81,18 @@ class CollectionSerializerTest extends AnyWordSpec with Matchers {
         val struct = MapStruct1(Map.empty)
         fury1.deserialize(fury1.serialize(struct)) shouldEqual struct
       }
+      "serialize/deserialize NestedMapStruct" in {
+        val struct = NestedMapStruct(Map("K1" -> Map("k1" -> "v1", "k2" -> "v2"), "K2" -> Map("k1" -> "v1")))
+        fury1.deserialize(fury1.serialize(struct)) shouldEqual struct
+      }
     }
   }}
 }
 
 case class CollectionStruct1(list: List[String])
 
+case class NestedCollectionStruct(list: List[List[String]], set: Set[Set[String]])
+
 case class MapStruct1(map: Map[String, String])
+
+case class NestedMapStruct(map: Map[String, Map[String, String]])

